File: backend/clients/snowflake_client.py
import asyncio
import snowflake.connector


class SnowflakeClient:
    def __init__(self, account: str, username: str, password: str, warehouse: str, database: str, schema: str):
        self.account = account
        self.username = username
        self.password = password
        self.warehouse = warehouse
        self.database = database
        self.schema = schema
        self._connection = None

    def _get_connection(self):
        if self._connection:
            return self._connection
        self._connection = snowflake.connector.connect(
            account=self.account,
            user=self.username,
            password=self.password,
            warehouse=self.warehouse,
            database=self.database,
            schema=self.schema,
        )
        return self._connection

    async def execute_query(self, query: str, binds: list | None = None) -> list[dict]:
        def _exec():
            conn = self._get_connection()
            cursor = conn.cursor(snowflake.connector.DictCursor)
            cursor.execute(query, binds or [])
            return cursor.fetchall()
        return await asyncio.to_thread(_exec)

    async def describe_table(self, table_name: str) -> list[dict]:
        rows = await self.execute_query(f"DESCRIBE TABLE {table_name}")
        return [
            {
                "name": row.get("name"),
                "type": row.get("type"),
                "nullable": row.get("null?") == "Y",
                "default": row.get("default"),
                "comment": row.get("comment"),
            }
            for row in rows
        ]

    def destroy(self):
        if self._connection:
            self._connection.close()
            self._connection = None

File: src/lib/clients/snowflake-client.ts
import snowflake from 'snowflake-sdk'

export class SnowflakeClient {
  private account: string
  private username: string
  private password: string
  private warehouse: string
  private database: string
  private schema: string
  private connection: any | null = null

  constructor(params: {
    account: string
    username: string
    password: string
    warehouse: string
    database: string
    schema: string
  }) {
    this.account = params.account
    this.username = params.username
    this.password = params.password
    this.warehouse = params.warehouse
    this.database = params.database
    this.schema = params.schema
  }

  private async getConnection(): Promise<any> {
    if (this.connection) return this.connection

    return new Promise((resolve, reject) => {
      const conn = snowflake.createConnection({
        account: this.account,
        username: this.username,
        password: this.password,
        warehouse: this.warehouse,
        database: this.database,
        schema: this.schema,
      })

      conn.connect((err: any) => {
        if (err) {
          reject(new Error(`Snowflake connection failed: ${err.message}`))
        } else {
          this.connection = conn
          resolve(conn)
        }
      })
    })
  }

  async executeQuery(query: string, binds?: any[]): Promise<any[]> {
    const conn = await this.getConnection()

    return new Promise((resolve, reject) => {
      conn.execute({
        sqlText: query,
        binds: binds || [],
        complete: (err: any, _stmt: any, rows: any[]) => {
          if (err) {
            reject(new Error(`Snowflake query failed: ${err.message}`))
          } else {
            resolve(rows || [])
          }
        },
      })
    })
  }

  async describeTable(tableName: string) {
    const rows = await this.executeQuery(`DESCRIBE TABLE ${tableName}`)
    return rows.map((row: any) => ({
      name: row.name,
      type: row.type,
      nullable: row['null?'] === 'Y',
      default: row.default,
      comment: row.comment,
    }))
  }

  destroy() {
    if (this.connection) {
      this.connection.destroy(() => {})
      this.connection = null
    }
  }
}
